L1Trigger/L1THGCal/plugins/be_algorithms/HGCalTriggerClusterIdentificationBDT.cc

```diff
@@ -0,0 +1,168 @@
+
+      private:
+        float pt_min_ = 0.;
+        float pt_max_ = std::numeric_limits<float>::max();
+        float eta_min_ = 1.5;
+        float eta_max_ = 3.;
+    };
+
+  public:
+
+    HGCalTriggerClusterIdentificationBDT();
+    ~HGCalTriggerClusterIdentificationBDT() override{};
+    void initialize(const edm::ParameterSet& conf) final;
+    float value(const l1t::HGCalMulticluster& cluster) const final;
+    bool decision(const l1t::HGCalMulticluster& cluster) const final;
+
+  private:
+    std::vector<Category> categories_;
+    std::vector<std::unique_ptr<TMVAEvaluator>> bdts_;
+    std::vector<double> working_points_;
+    std::vector<std::string> input_variables_;
+
+    float clusterVariable(const std::string&, const l1t::HGCalMulticluster&) const;
+    int category(float pt, float eta) const;
+
+
+};
+
+DEFINE_EDM_PLUGIN(HGCalTriggerClusterIdentificationFactory,
+    HGCalTriggerClusterIdentificationBDT,
+    "HGCalTriggerClusterIdentificationBDT" );
+
+
+HGCalTriggerClusterIdentificationBDT::
+HGCalTriggerClusterIdentificationBDT():HGCalTriggerClusterIdentificationBase()
+{
+}
+
+void
+HGCalTriggerClusterIdentificationBDT::
+initialize(const edm::ParameterSet& conf)
+{
+  categories_.clear();
+  bdts_.clear();
+  working_points_.clear();
+  input_variables_.clear();
+  input_variables_ = conf.getParameter< std::vector<std::string> >("Inputs");
+  std::vector<std::string> bdt_files = conf.getParameter< std::vector<std::string> >("Weights");
+  std::vector<double> categories_etamin = conf.getParameter<std::vector<double>>("CategoriesEtaMin");
+  std::vector<double> categories_etamax = conf.getParameter<std::vector<double>>("CategoriesEtaMax");
+  std::vector<double> categories_ptmin = conf.getParameter<std::vector<double>>("CategoriesPtMin");
+  std::vector<double> categories_ptmax = conf.getParameter<std::vector<double>>("CategoriesPtMax");
+  working_points_ = conf.getParameter<std::vector<double>>("WorkingPoints");
+  for(unsigned cat=0; cat<categories_etamin.size(); cat++)
+  {
+    categories_.emplace_back(
+        categories_ptmin[cat],
+        categories_ptmax[cat],
+        categories_etamin[cat],
+        categories_etamax[cat]);
+  }
+  std::vector<std::string> spectators = {};
+  for (const auto& file : bdt_files)
+  {
+    bdts_.emplace_back(new TMVAEvaluator());
+    bdts_.back()->initialize(
+        "!Color:Silent:!Error",
+        "BDT::bdt",
+        edm::FileInPath(file).fullPath(),
+        input_variables_,
+        spectators,
+        false, false);
+  }
+}
+
+float
+HGCalTriggerClusterIdentificationBDT::
+value(const l1t::HGCalMulticluster& cluster) const
+{
+  std::map<std::string, float> inputs;
+  for(const std::string& var : input_variables_)
+  {
+    inputs[var] = clusterVariable(var, cluster);
+  }
+  float pt = cluster.pt();
+  float eta = cluster.eta();
+  int cat = category(pt, eta);
+  return (cat!=-1 ? bdts_.at(cat)->evaluate(inputs) : -999.);
```

Comment:
we might want to consider adding/modifying interfaces for `TMVAEvaluator`, creating a map is probably more work than necessary. (outside the scope of this PR, but just so I remember later)